### PR TITLE
testing: Avoid using context.Background

### PR DESCRIPTION
**Pull request by gauravgahlot**
Fixes: #1878

### CODE REVIEW

**Comment by gauravgahlot** [comment]
I have intentionally split the changes into commits based on the parent directory. This will make it easier for me to do any changes suggested during the review. 

@dfawley Could you please add `no release notes` label and also take a look at the changes?

**Review by easwars — CHANGES_REQUESTED** [review]
We don't want to change `context.Background()` with `context.TODO()`. What we want instead is what @dfawley mentioned in https://github.com/grpc/grpc-go/issues/1878#issuecomment-702970169.

Each test file (one per package) should define a `defaultTestTimeout`. And each test must create a local context with the appropriate timeout.

```
const defaultTestTimeout = 10 * time.Second

func TestXxx(t *testing.T) {
  ctx, cancel := context.WithTimeout(defaultTestTimeout)
  defer cancel()

  // Do stuff and pass `ctx` to functions which accept one.
}
```

And for places where we block on some event *not* happening, we should define a much shorter timeout (something of the order of 10ms), create a new context with that short deadline, and use that.

Hope this helps.

Thanks.

**Comment by gauravgahlot** [comment]
Thanks for the explanation @easwars. I clearly misinterpreted @dfawley's comment. 

I have made changes to the last commit. Please take a look and let me know if I'm doing it right this time.
If so, I will update the rest of the commits.

Thanks again. :)

File: benchmark/primitives/context_test.go [hunk 1]
```diff
@@ -139,7 +141,8 @@ func BenchmarkContextWithValue(b *testing.B) {
 		{"newContextWithGlobalKey", newContextWithGlobalKey},
 	}
 
-	pCtx := context.Background()
+	pCtx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
```
Comment (easwars) [inline benchmark/primitives/context_test.go:144]: Here, we are testing `context.WithValue()`, so it is totally OK to use the background context.

What you can fix here though are the `BenchmarkTimer*` functions. Those are using arbitrary large values, those should be changed use the `defaultTestTimeout`.

File: metadata/metadata_test.go [hunk 2]
```diff
@@ -242,15 +251,17 @@ func Benchmark_AddingMetadata_ContextManipulationApproach(b *testing.B) {
 func BenchmarkAppendToOutgoingContext(b *testing.B) {
 	const num = 10
 	for n := 0; n < b.N; n++ {
-		ctx := context.Background()
+		ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
```
Comment (easwars) [inline metadata/metadata_test.go:254]: This should probably be moved to be outside the `for` loop. Thanks.

File: stats/stats_test.go [hunk 3]
```diff
@@ -281,7 +283,10 @@ func (te *test) doUnaryCall(c *rpcConfig) (*testpb.SimpleRequest, *testpb.Simple
 	} else {
 		req = &testpb.SimpleRequest{Id: errorID}
 	}
-	ctx := metadata.NewOutgoingContext(context.Background(), testMetadata)
+
+	tCtx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	ctx := metadata.NewOutgoingContext(tCtx, testMetadata)
```
Comment (easwars) [inline stats/stats_test.go:289]: This call to `metadata.NewOutgoingContext()` could be inlined within the `UnaryCall`, just like how it is done for the `FullDuplexCall`.

**Review by easwars — CHANGES_REQUESTED** [review]
I checked the last commit. The changes look mostly OK. I left some minor comments. Thank you so very much for doing this.

File: test/end2end_test.go [hunk 4]
```diff
@@ -1263,14 +1269,19 @@ func testConcurrentServerStopAndGoAway(t *testing.T, e env) {
...
 	tc := testpb.NewTestServiceClient(cc)
-	stream, err := tc.FullDuplexCall(context.Background(), grpc.WaitForReady(true))
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	stream, err := tc.FullDuplexCall(ctx, grpc.WaitForReady(true))
 	if err != nil {
 		t.Fatalf("%v.FullDuplexCall(_) = _, %v, want <nil>", tc, err)
 	}
+	cancel()
+
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
```
Comment (easwars) [inline test/end2end_test.go:1279]: Can't we use the `ctx` created in line 1272 here? And do a `defer cancel()`

**Review by easwars — APPROVED** [review]
Thank you so very much.

**Comment by easwars** [comment]
@gauravgahlot : Could you please take care of the test failure here: https://github.com/grpc/grpc-go/pull/3949/checks?check_run_id=1340435962. Thanks.

**Comment by gauravgahlot** [comment]
@easwars I have fixed the issue and it should be good to go now. 🚀

**Pull request merged**